AstSemantics.md

```diff
@@ -273,39 +275,53 @@ nested. This guarantees that all resulting control flow graphs are well-structur
 
 ## Calls
 
-Direct calls to a function specify the callee by index into a function table.
+Each function has a *signature*, which consists of:
 
-  * `call_direct`: call function directly
+  * Return types, which are a sequence of local types
+  * Argument types, which are a sequence of local types
+
+Note that WebAssembly itself does not support variable-length argument lists.
+C and C++ compilers are expected to implement this functionality by storing
+an array of arguments in a buffer in linear memory and passing a pointer to
+the buffer.
 
-Each function has a signature in terms of expression types, and calls must match
-the function signature
-exactly. [Imported functions](MVP.md#code-loading-and-imports) also have
-signatures and are added to the same function table and are thus also callable
-via `call_direct`.
+In the MVP, the length of the return types vector may only be 0 or 1. This
+restriction may be lifted in the future with the addition of support for
+[multiple return values](FutureFeatures.md#multiple-return-values").
 
-Indirect calls may be made to a value of function-pointer type. A function-
-pointer value may be obtained for a given function as specified by its index
-in the function table.
+There are two forms of calls:
 
+  * `call_direct`: call function directly
   * `call_indirect`: call function indirectly
-  * `addressof`: obtain a function pointer value for a given function
 
-Function-pointer values are comparable for equality and the `addressof` operator
-is monomorphic. Function-pointer values can be explicitly coerced to and from
-integers (which, in particular, is necessary when loading/storing to memory
-since memory only provides integer types). For security and safety reasons,
-the integer value of a coerced function-pointer value is an abstract index and
-does not reveal the actual machine code address of the target function.
+Direct calls identify their function statically. Indirect calls have a
+funcptr operand which identifies the function at runtime.
+
+Calls have a signature, which is the expected return types and argument types
+(ignoring the funcptr operand, in the case of `call_indirect`) of the
+AST node. Call operations trap if the signature of the call differs from the
+signature of the called function.
 
-In the MVP, function pointer values are local to a single module. The
+### Function pointers
+
+Function pointer values are obtained through the use of a special operator:
+
+  * `addressof`: obtain a funcptr value for a given statically-identified function
```

Comment:
I think we should avoid using the name ```funcptr``` because it implies that it's a pointer into the heap, when it's actually an index into a table somewhere (and there are probably separate tables per signature type). ```funcid```? Not really interested in bikeshedding beyond that implying pointer-ness is probably bad here, even if compilers happen to use this type to *implement* function pointers.